New objective functions and documentation

samgdotson opened the pull request:
### Please review PR #38 before reviewing this one!

## Updates

This PR adds several things
* generic objectives (e.g., `per_unit_energy` or `per_unit_capacity`)
* a new objective that measures volatility
* documentation for these objectives and examples for their use
* tests for these objectives

and it updates the `capacity_expansion` module to account for generic objectives.

samgdotson committed:
adds a new objective function

samgdotson committed:
Merge branch 'techsub_revamp' into wpe

samgdotson committed:
Merge branch 'techsub_revamp' of github.com:samgdotson/osier into wpe

samgdotson committed:
adds general objective functions, equations, and documentation.

samgdotson committed:
Merge branch 'techsub_revamp' of github.com:samgdotson/osier into wpe

samgdotson committed:
pep8 fixes

abachma2 commented on osier/equations.py:
```diff
@@ -78,14 +80,113 @@ def get_dispatchable_names(technology_list):
+    ...                    functools.partial(per_unit_capacity, attribute='employment')]
+
+    """
+    objective_value = np.array([getattr(t, attribute) * t.capacity
+                                for t in technology_list
+                                if hasattr(t, attribute)]).sum()
```
If you want things per unit capacity, then wouldn't you divide by the capacity?

samgdotson replied:
No, this refers to certain quantities that are measured in a per-unit-capacity basis. These include things like land use or employment. But there could be others that I am not considering, thus I created a general function to handle a user-defined attribute and corresponding objective without editing the source code. 

The *attribute* is per-unit-capacity. The objective is the *total* of that attribute.

abachma2 replied:
Then I feel that the method name is misleading. I thought this was to convert from a total to a per unit basis.

samgdotson replied:
Well, it's referring to a class of objectives (same with per_unit_energy). Could you suggest an alternative name?

@LukeSeifert @yardasol @nsryan2 what do you think?

LukeSeifert replied:
Maybe `objective_capacity` and `objective_energy`?

samgdotson replied:
I think those suggest that capacity and energy are the things being calculated.

abachma2 replied:
Maybe `total_objective_capacity` and `total_objective_energy`.

nsryan2 replied:
Normally I'm pro-simple variable names, but, if you are going to calculate capacity or energy somewhere else, I think Amanda's suggestion is very clear

samgdotson replied:
Again, I think that suggests the total capacity and total energy are being calculated -- which are not objectives. A per_unit_energy objective would be something like carbon emissions which are only emitted when energy is produced.

yardasol replied:
Maybe `calculate_capacity_objective` and `calculate_energy_objective`?

LukeSeifert replied:
Possibly combine the functions into `generate_objective` and add a parameter to choose between capacity and energy?

samgdotson committed:
adds kwargs to dispatch model, reorders optional arguments

samgdotson committed:
adds kwargs to dispatch model, reorders optional arguments

abachma2 commented on osier/equations.py:
```diff
@@ -78,14 +80,113 @@ def get_dispatchable_names(technology_list):
+                           for t in dispatch_techs + non_dispatch_techs
+                           if hasattr(t, attribute)])
+
+    objective_value = np.dot(
+        attributes,
+        dispatch_results[column_names].values.T).sum()
```
Does this include the energy somewhere? It's not clear to me, but I'm guessing it's in the attributes object?

samgdotson replied:
The energy produced by each technology is determined by the dispatch model -- so the energy is in `dispatch_results`. The column names correspond to technology names. There are other columns such as `LoadLoss`, `Battery_level`, `cost`, which are important results to track but unrelated to the energy produced.

abachma2 replied:
That makes sense. I wasn't well-versed on the data structures you use.

abachma2 commented on osier/equations.py:
```diff
@@ -143,32 +247,34 @@ def annual_co2(technology_list, solved_dispatch_model, emission='co2'):
-                         for t in dispatch_techs+non_dispatch_techs 
-                         if hasattr(t,emission)])
+    emissions = np.array([getattr(t, emission)
+                         for t in dispatch_techs + non_dispatch_techs
+                         if hasattr(t, emission)])
 
-    emissions_total = np.dot(emissions, dispatch_results[column_names].values.T).sum()
+    emissions_total = np.dot(
+        emissions,
+        dispatch_results[column_names].values.T).sum()
```
This is really similar to what you do in `per_unit_energy`. Would it make sense to create a method to do this?

samgdotson replied:
This is a good point, now that I have the generic functions, I should be calling those for each unique objective.

samgdotson replied:
I'm realizing that might not have been what you meant. Why would I have an emissions calculator when `per_unit_energy` can also do that? Usability. In order for users to use `per_unit_energy` they have to do

```py
import functools
objectives = [functools.partial(per_unit_energy, attribute='co2_rate')]
```
While straightforward, it's not as readable as 

```py
objectives = [annual_emission]
```

I'll keep adding functions to `equations.py` because it's always clearer to have a specific function. The generic version adds functionality but sacrifices usability.

abachma2 reviewed (commented):
Hi @samgdotson. Great work on this PR. I have a few comments about some things. Many of the parts in this PR felt repeated from what was in #38. Should that be the case?

samgdotson committed:
updates doc string in technology.py

LukeSeifert commented on osier/equations.py:
```diff
@@ -78,14 +80,113 @@ def get_dispatchable_names(technology_list):
     return dispatchable_names
 
 
+def per_unit_capacity(technology_list,
```
It doesn't look like `solved_dispatch_model` is used in this function

samgdotson replied:
It's not used, but the uniformity of objective function structures allow users to define their own objectives and pass them to osier. Otherwise, I would have to devise a way to check what kind of objective is being
passed (i.e. whether it needs a solved model or not) so that osier could give it the right information.

For concreteness, during a genetic algorithm simulation the following occurs

1. Osier generates a test population of technology capacities.
2. Those technologies are passed to the dispatch model which determines how much energy each one produces
3. It calculates the user-specified objectives.

```py
from osier import CapacityExpansion
from osier import total_cost, annual_emission
objectives_list = [total_cost, annual_emission]

problem = CapacityExpansion(technology_list,
    demand,
    objectives=objectives_list)
###############################3
# inside the model run
# generate capacities
# determine dispatch

for obj_func in self.objectives:
    obj = obj_func(technology_list=self.technology_list,
        solved_dispatch_model=self.dispatch_model)
    out['F'].append(obj)
```

For more context, look at [CapacityExpansion lines 176-188](https://github.com/arfc/osier/blob/da88fd6418afed238b467fd918f8f425e8e58144/osier/models/capacity_expansion.py#L176-L188).

LukeSeifert commented on osier/equations.py:
```diff
@@ -78,14 +80,113 @@ def get_dispatchable_names(technology_list):
+    return objective_value
+
+
 def annualized_capital_cost(technology_list, solved_dispatch_model=None):
```
This function also calls for the `solved_dispatch_model` but doesn't look like it uses it?

samgdotson replied:
See my previous comment! This uniformity is part of the "osier style."

LukeSeifert commented on osier/equations.py:
```diff
@@ -101,15 +202,15 @@ def annualized_capital_cost(technology_list, solved_dispatch_model=None):
         The annual capital cost of the technology set.
     """
-    capital_cost = np.array([t.total_capital_cost / t.lifetime 
+    capital_cost = np.array([t.total_capital_cost / t.lifetime
```
Does this need to be discounted since it is turned into an annual capital cost?

samgdotson replied:
At the moment, the "model time horizon" consists of one year. Perhaps in future iterations the discounting will matter, but right now this simplification is appropriate.

LukeSeifert commented on osier/equations.py:
```diff
@@ -190,3 +296,50 @@ def total_cost(technology_list, solved_dispatch_model):
 
     return total_cost
```
This is the total cost per year, right?

samgdotson replied:
correct

LukeSeifert commented on osier/models/capacity_expansion.py:
```diff
@@ -52,52 +45,69 @@ class CapacityExpansion(ElementwiseProblem):
+                 penalty=LARGE_NUMBER,
+                 power_units=u.MW,
+                 curtailment=True,
+                 allow_blackout=False,
```
Previously said default was True

samgdotson replied:
I go back and forth on this. Mostly because I want to limit the number of infeasible technology mixes that osier tries during optimization. In this sense, leaving it True works as intended, but the objective cost is far too dominant. I'm pursuing other methods to improve computational efficiency.

LukeSeifert replied:
I think whichever default you think works better should be used in the docstring to keep things consistent. If you end up switching it back to `True` by default later, you can update the docstring to match the change.

samgdotson committed:
Merge branch 'main' of github.com:arfc/osier into wpe

samgdotson committed:
calls per_unit_energy in the annual emission function

samgdotson commented:
Thanks for the review @abachma2. There might be some repetition since I was working on this and the previous PR simultaneously. Fortunately there are no code conflicts (per github). I pulled in the previous merge so it should be okay.

LukeSeifert commented on osier/models/capacity_expansion.py:
```diff
@@ -71,39 +63,39 @@ class CapacityExpansion(ElementwiseProblem):
     allow_blackout : boolean
         If True, a "reliability" technology is added to the dispatch model that will
         fulfill the mismatch in supply and demand. This reliability technology
-        has a variable cost of 1e4 $/MWh. The value must be higher than the 
-        variable cost of any other technology to prevent a pathological 
+        has a variable cost of 1e4 $/MWh. The value must be higher than the
+        variable cost of any other technology to prevent a pathological
         preference for blackouts. Default is True.
```
If `allow_blackout` is true, then it will fulfill the mismatch in supply and demand? I think it would be more intuitive if this happened if `allow_blackout` was false.

samgdotson replied:
Since this dispatch model is an LP problem, demand = supply at every time step. Otherwise, the model is infeasible. If I want to model situations where demand cannot be met (i.e. "allowing blackouts" to occur), I add a dummy technology  whose fuel cost corresponds to the "value of lost load." I can add a separate section in the docs to explain this if that would make you happy! (If this is what you want to see, please make an issue!)

LukeSeifert replied:
I think I understand a bit better. So if `allow_blackout` is `True`, you add a "reliability" technology to capture the value of the lost load. If `allow_blackout` is `False`, the "reliability" technology isn't included. In that case, is demand limited or is supply raised to prevent a blackout from occurring, or does supply not meet demand in that case?

samgdotson replied:
In that case the model is infeasible because supply cannot meet demand.

LukeSeifert reviewed (commented):
Overall looks good. I just had one or two things that I had questions on.

LukeSeifert reviewed (approved):
Thanks for answering my questions! I have two minor points still up, but they are minor things.

samgdotson committed:
updates doc string for allow_blackout default

samgdotson commented:
Thanks for the review @LukeSeifert, if you're satisfied you can go ahead and merge. I updated the doc string to reflect the change in the default value.

samgdotson committed:
updates doc strings, documentation, and moves function locations

samgdotson committed:
rename ambigious functions

samgdotson committed:
propagates function renaming to tests

samgdotson commented:
@abachma2 @LukeSeifert @yardasol @munkm I renamed the two ambiguous functions `objective_from_X`. Hopefully, this satisfies everyone. I also added equations to the docstrings.

samgdotson commented:
Nudge @abachma2 @yardasol

abachma2 reviewed (approved):
Hi @samgdotson. Thank you for adjusting those function names. I'm not sure if they are the best they can be, but they are definitely less confusing than they previously were. I will make an issue to remind us to keep thinking of any better names. I am approving this PR, but will let someone else merge, since multiple people have looked at this PR.
